Document how to localized code strings of type LocalizedStringResource

mman opened the pull request:
Fixes https://github.com/FlineDev/BartyCrouch/issues/265

## Proposed Changes

  - When no `update.code.customFunction` is specified, use `LocalizedStringResource` by default, in addition to `NSLocalizedString`, and `CFCopyLocalizedString`.
  - Document how to use `LocalizedStringResource(_: StaticString, comment: StaticString)` initializer to localize SwiftUI and AppIntents code.

## Alternatives Considered

When examining manpage for `genstrings(1)` and help obtained via `xcrun extractLocStrings`, both commands support a `-SwiftUI` argument that should parse SwiftUI `Text()` initializer. Therefore it may appear useful to expose `-SwiftUI` argument via `update.code` configuration.

I have not used the `-SwiftUI` option and do not know how reliable or useful is/would be in this case. And I am not sure what other instances apart from `Text()` will be recognized. Perhaps the `xcrun extractLocStrings` may actually analyze the `.swift` files and precompile them to recognize all use of `LocalizedStringResource` and extract those as well. Needs more investigation.

Jeehut commented on README.md:
```diff
@@ -258,7 +258,7 @@ tasks = ["interfaces", "code", "transform", "normalize"]
 - `defaultToKeys`: Add new keys both as key and value.
 - `additive`: Prevents cleaning up keys not found in code.
-- `customFunction`: Use alternative name to `NSLocalizedString`.
+- `customFunction`: Use alternative name to search for strings to localize, in addition to `NSLocalizedString`, and `CFCopyLocalizedString`. If not specified, look for `LocalizedStringResource`.
```
This could a bit shorter. For example:
```suggestion
- `customFunction`: Use alternative name to search for strings to localize, in addition to `NSLocalizedString`, and `CFCopyLocalizedString`. Defaults to `LocalizedStringResource`.
```

Jeehut commented on README.md:
```diff
@@ -355,6 +355,40 @@ self.title = L10n.Onboarding.FirstPage.headerTitle
+
+Historically, Apple platforms used `CFCopyLocalizedString`, and `NSLocalizedString` macros and their variants, to mark strings used in code to be localized, and to load their localized versions during runtime from `Localizable.strings` file.
+
+Since introduction of SwiftUI, the localized strings in code are typically typed as `LocalizedStringResource`, and are no longer marked explicitely.
```
Aren't SwiftUI views localized via `LocalizedStringKey`? As far as I know, only few frameworks make use of `LocalizedStringResource` and unlike `LocalizedStringKey`, the `LocalizedStringResource` type isn't defined in the SwiftUI framework but [in Foundation](https://developer.apple.com/documentation/foundation/localizedstringresource). Therefore I find this sentence misleading, but maybe I'm missing something. I would change as follows:
```suggestion
Since introduction of the AppIntents framework, the localized strings in code can also be typed as `LocalizedStringResource`, and are no longer marked explicitly.
```

Jeehut commented on README.md:
```diff
@@ -355,6 +355,40 @@ self.title = L10n.Onboarding.FirstPage.headerTitle
 
 NOTE: As of version 4.x of BartyCrouch *formatted* localized Strings are not supported by this automatic feature.
 
+### Localizing strings of `LocalizableStringResource` type (AppIntents, SwiftUI, ...)
```
Note sure if `LocalizedStringResource` is used in SwiftUI, you might mix it up with `LocalizedStringKey`...

Jeehut reviewed (changes requested):
The code changes look good to me, but I left some comments in the documentation. There might be a misunderstanding here (either on my side or yours), please check and let's clarify.

mman commented:
@Jeehut You are absolutely right about me being absolutely wrong about SwiftUI localization. My bad! I have addressed your suggestions. The PR should be ready now. But it brings me to question that probably belongs someplace else, and that is. What about SwiftUI localization with `bartycrouch` then? How would that work? I will try to investigate... just dropping it here...

mman commented:
@Jeehut The swift ui localization question meaning: `xcrun extractLocStrings` does contain the `-SwiftUI` argument, as well as `genstrings` does have it. But the documentation is sparse, saying it only supports extraction of `Text()`. I am not familiar with SwiftUI that deep, but logical way forward would be to expose the `-SwiftUI` argument into the config. Correct?

Jeehut commented:
This looks good to me, merging. You just forget to add a changelog entry, but I'll add one before release then.

Jeehut commented:
> What about SwiftUI localization with bartycrouch then? How would that work?

SwiftUI totally supports plain old translation methods such as NSLocalizedString etc. as these end up in a `String` which all SwiftUI views accept as an alternative to `LocalizedStringKey`, so general SwiftUI support is already there. Having that said, only if `LocalizedStringKey` is used, one can have SwiftUI previews with multiple languages of the same view side-by-side to ensure everything looks nice. Note that [ReMafoX](https://remafox.app) fully supports `LocalizedStringKey`s as well. It all comes down to the SwiftGen template being used, ReMafoX has its own Swift enum generation logic, but one could also build a custom SwiftGen template where `LocalizedStringKey` is supported if really needed. This has nothing to do with BartyCrouch directly.

mman commented:
Thanks!
